ManageIQ/manageiq
# Fix for UI hangs if errors are encountered during Git Import

The main issue is that if github (or whatever site they're trying to use for an import) is down, it takes a really long time to timeout. @mkanoor and I figured out that in production, the apache timeout is 2 minutes, as is the timeout for the backend to decide that it should give up on trying to refresh the repo. This, coupled with the queue latency causes a 302 redirect to come back from the rails server, but apache instead serves up a 502 proxy error.

This fix aims to address this issue by instead not waiting for the task to finish on the server side, but on the client side. When we queue the task, we immediately pass the task id back to the client, which uses that to poll the server for completeness. I've currently set the poll to 1.5 seconds, if you think it should be longer or shorter, please let me know.

Links
----------------
https://bugzilla.redhat.com/show_bug.cgi?id=1393982

Steps for Testing/QA
-------------------------------
To mimic this in development, you can either add something to /etc/hosts like `8.8.8.8 github.com` and continue using github.com. Or, when importing, just use a local IP that doesn't resolve to anything in the url field, e.g. `https://192.168.1.42/test_org/test_repo`.

@miq-bot add_label blocker, bug

@miq-bot assign @gmcculloug 

@mkanoor Can you test and make sure everything appears to be working the same as it was before? I ran through a bunch of scenarios and it seemed to be behaving nicely, but I'd like a sanity check.
@syncrou Can you review as there's a lot of javascript changes I made here that I'd like your eye on.
